src/audio/peak_volume.c

```diff
@@ -0,0 +1,440 @@
+
+static int init_peakvol(struct comp_dev *dev, struct comp_ipc_config *config,
+			void *spec, enum ipc4_vol_mode mode)
+{
+	struct ipc4_peak_volume_module_cfg *peakvol = spec;
+	struct peakvol_data *cd;
+	uint8_t channel_cfg;
+	uint8_t channel;
+
+	dev->ipc_config = *config;
+	list_init(&dev->bsource_list);
+	list_init(&dev->bsink_list);
+
+	dcache_invalidate_region(spec, sizeof(*peakvol));
+	cd = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(cd->base));
+	if (!cd) {
+		comp_free(dev);
+		return -ENOMEM;
+	}
+
+	mailbox_hostbox_read(&cd->base, sizeof(cd->base), 0, sizeof(cd->base));
+	cd->active_channels = cd->base.audio_fmt.channels_count;
+	cd->mode = mode;
+
+	for (channel = 0; channel < cd->active_channels; channel++) {
+		if (peakvol->config[0].channel_id == IPC4_ALL_CHANNELS_MASK)
+			channel_cfg = 0;
+		else
+			channel_cfg = channel;
+
+		cd->peak_regs.current_volume_[channel] = 0;
+		set_volume(cd, channel, peakvol->config[channel_cfg].target_volume,
+			   peakvol->config[channel_cfg].curve_type,
+			   peakvol->config[channel_cfg].curve_duration);
+		cd->peak_regs.peak_meter_[channel] = 0;
+	}
+
+	comp_set_drvdata(dev, cd);
+
+	return 0;
+}
+
+static struct comp_dev *peakvol_new(const struct comp_driver *drv,
+				    struct comp_ipc_config *config,
+				    void *spec)
+{
+	struct peakvol_data *cd;
+	struct comp_dev *dev;
+	int ret;
+
+	comp_cl_dbg(&comp_peakvol, "peakvol_new()");
+
+	dev = comp_alloc(drv, sizeof(*dev));
+	if (!dev)
+		return NULL;
+
+	ret = init_peakvol(dev, config, spec, IPC4_MODE_PEAKVOL);
+	if (ret < 0) {
+		comp_free(dev);
+		return NULL;
+	}
+
+	cd = comp_get_drvdata(dev);
+	cd->mailbox_offset = offsetof(struct ipc4_fw_registers, peak_vol_regs);
+	cd->mailbox_offset += IPC4_INST_ID(config->id) * sizeof(struct ipc4_peak_volume_regs);
+
+	dev->state = COMP_STATE_READY;
+	return dev;
+}
+
+static struct comp_dev *gain_new(const struct comp_driver *drv, struct comp_ipc_config *config,
+				 void *spec)
+{
+	struct comp_dev *dev;
+	int ret;
+
+	comp_cl_dbg(&comp_gain, "gain_new()");
+
+	dev = comp_alloc(drv, sizeof(*dev));
+	if (!dev)
+		return NULL;
+
+	ret = init_peakvol(dev, config, spec, IPC4_MODE_GAIN);
+	if (ret < 0) {
+		comp_free(dev);
+		return NULL;
+	}
+
+	dev->state = COMP_STATE_READY;
+	return dev;
+}
+
+static void peakvol_free(struct comp_dev *dev)
+{
+	struct peakvol_data *cd = comp_get_drvdata(dev);
+	struct ipc4_peak_volume_regs regs;
+
+	/* clear mailbox */
+	memset_s(&regs, sizeof(regs), 0, sizeof(regs));
+	mailbox_sw_regs_write(cd->mailbox_offset, &regs, sizeof(regs));
+
+	rfree(cd);
+	rfree(dev);
+}
+
+static void gain_free(struct comp_dev *dev)
+{
+	struct peakvol_data *cd = comp_get_drvdata(dev);
+
+	rfree(cd);
+	rfree(dev);
+}
+
+static inline void update_peakvol_in_mailbox(struct peakvol_data *cd)
+{
+	mailbox_sw_regs_write(cd->mailbox_offset, &cd->peak_regs, sizeof(cd->peak_regs));
+}
+
+static void peakvol_process_sample_16bit(struct comp_dev *dev, struct audio_stream *sink,
```

Comment:
How is this different from volume.c ? We need to add the features to volume.c not create a new volume.

Reply:
difference: (1) volume format is different.  IPC4 uses format 1.31, but sof uses 8.16 (2) IPC4 caculate max volume value of stream sample and update it to mailbox. And we need to adopt windows  style fade feature. the algorithm is difference with current volume component. Adding feature to volume will lead to totally rework volume and make it hard to understand

Reply:
> difference: (1) volume format is different. IPC4 uses format 1.31, but sof uses 8.16 

This is fine, different formats are supported today in pipelines, this can be a buffer attribute but @singalsu needs to comment here on changes needed for 1.31 (and whether it provides similar quality).

> (2) IPC4 caculate max volume value of stream sample and update it to mailbox. 

This action can be controlled by a flag

> (3) we need to adopt windows style fade feature. the algorithm is difference with current volume component. Adding feature to volume will lead to totally rework and make it hard to understand

Not really, fade is controlled by a flag and can be checked on every copy(). @singalsu will make sure it aligns with requirements.

Reply:
The Q1.31 provides max gain of 0 dB while SOF can support up to +40 dB. We allow > 0 dB microphone gains in several topologies for capture. We can change volume to Q1.31 but need to address the gain boost with other component, e.g. EQ. Some > 0 dB gains might also be as default in UCM so need to check those as well if the feature has been used. One less critical issue is that Q1.31 do not allow exactly 1.0 gain, so we may not be able to achieve 1:1 PCM copy trough volume.

Reply:
0x7fffffff is for exactly 1.0 gain. this means 0x7fffffff is not used for normal gain